benchadapt/python/benchadapt/adapters/_adapter.py

```diff
@@ -17,16 +17,42 @@ class BenchmarkAdapter(abc.ABC):
     command : List[str]
         A list of args to be run on the command line, as would be passed
         to `subprocess.run()`.
+    result_defaults_override : Dict[str, Any]
+        A dict of default values to be passed to `BenchmarkResult`. Useful for
+        specifying metadata only available at runtime, e.g. build info. Overrides
+        `BenchmarkResult` defaults; is overridden by values passed in
```

Comment:
One thing I'm confused about is that someone implementing an adapter needs to know to always call `self.curried_benchmark_result()` on each benchmark result before returning them from `self.transform_results()`. Could we automatically do that in `run()` instead?

Reply:
Yeah, they need use use the method instead of the class. (Well technically they don't have to, but if they use `BenchmarkResult` directly users won't be able to override anything.) I don't love this idiom, but I don't have a better idea at the moment.

Reply:
That would only work for changing fields after-the-fact, not before, which is useful to prevent fields from getting auto-populated, e.g. if we specify `cluster_info`, `machine_info` will stay `None` (as it should so it can get posted) instead of getting auto-populated.

And we can't necessarily pass in a partially pre-populated instance, because some fields like `timestamp` we really do want to be generated at runtime (though this probably doesn't matter too much). 

But I think you might be right; the path forward here might be only editing after the fact, and improving `BenchmarkResult` so setting `cluster_info` will unset `machine_info` (which probably involves moving it out of `__post_init__` and figuring out how to make dataclasses and properties play nice...lordy, it's attrs all over again). I'll play a little and see what I can do.

Reply:
Ok, adjusted `BenchmarkResult` so we can run everything after-the-fact. Had to add a little wackiness (there's now `cluster_info`, `_cluster_info`, and `__cluster_info` 😬), but the `__post_init__` is now gone and post-init setting of `cluster_info` is now handled properly. I moved the abstract method from `transform_results()` to `_transform_results()`; the former now calls the latter plus updates the results with any metadata set on init. 

Much happier with this version; there's still some hackiness, but writers of adapters don't have to worry about it, and users of adapters still get all the functionality.